sprint_14

Fennec-meow opened the pull request.

Fennec-meow committed:
first commit

Fennec-meow committed:
sprint_14

The Tracker application model has been created

1. The following controllers were added:

a. TrackersViewController - The main controller;
b. CreatingTrackerViewController - The transition controller for creating "Habits" and "Irregular events";
c. CreatingHabitViewController - The controller for creating Habits" and "Irregular events";
d. CategoryViewController - The controller for selecting "Categories";
e. NewCategoryViewController - The controller to create a new "Category";
f. ScheduleViewController - The controller for selecting the "Schedule".

2. The creation of cells has been implemented:

a. CustomTableViewCell - Shows the contents of the table on the "CreatingHabitViewController" and "ScheduleViewController" screens;
b. TrackerCollectionViewCell - Shows the collection cell on the "TrackersViewController" screen;
c. SupplementaryView - Shows the title of the created cell "TrackerCollectionViewCell" in "TrackersViewController".

3. The following entities have been created:

a. struct Tracker is an entity for storing information about a tracker (for a "Habit" or "Irregular Event");
b. struct TrackerCategory is an entity for storing trackers by category.
c. struct TrackerRecord is an entity for storing a record that a certain tracker was completed on a certain date;
d. enum TypeTrackers is an entity for determining a "Habit" or "Irregular Event";
e. enum WeekDay is an entity for creating a "Schedule".

4. Added a class that stores static data:

a. DataManager - Contains a ready-made list of "Categories" and "Trackers".

Fennec-meow committed:
first commit

NaR0RoW commented on Tracker.xcodeproj/xcuserdata/mac.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist:
Папку `xcuserdata` стоит игнорировать, кладя в `.gitignore`. В ней хранится информация о состоянии проекта (какой файл открыт, какой таргет выбран и т.п.), брейкпоинты и схемы для текущего пользователя `Xcode`.
Про `gitignore` можно прочитать вот [тут](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files)

NaR0RoW commented on Tracker/AppDelegate.swift:
```diff
@@ -9,14 +9,18 @@ import UIKit
 
 @main
 class AppDelegate: UIResponder, UIApplicationDelegate {
```
Стоит добавить модификатор `final`, чтобы вызовы методов работали быстрее.
Но это уже со звездочкой, хотя на собеседованиях спрашивают. Можно почитать про [диспачеризацию тут](https://habr.com/ru/articles/673400/)

NaR0RoW commented on Tracker/Resources/Assets.xcassets/Loader/Loader.imageset/Loader.png:
Картинки 1x лучше не добавлять. Они нужны были для iPhone без retina экранов, а последний такой - iPhone 4. Он не поддерживает минимальную версию приложения, поэтому можно смело удалять, чтобы не занимать лишнюю память

NaR0RoW commented on Tracker/SceneDelegate.swift:
```diff
@@ -13,11 +13,13 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+    
+
 
     func sceneDidDisconnect(_ scene: UIScene) {
```
Можно удалить не используемые методы

NaR0RoW commented on Tracker/SceneDelegate.swift:
```diff
@@ -13,11 +13,13 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
-        // If using a storyboard, the `window` property will automatically be initialized and attached to the scene.
-        // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
-        guard let _ = (scene as? UIWindowScene) else { return }
+        guard let windowScene = (scene as? UIWindowScene) else { return }
```
Лучше не использовать круглые скобки, если это явно не требуется

NaR0RoW commented on Tracker/StartViewController/StartViewController.swift:
```diff
@@ -0,0 +1,140 @@
+    // MARK: ImageConstants
+    
+    enum ImageConstants {
+        static let backgroundImage = UIImage(named: "backgroundBlueImage")
```
Лучше использовать `UIImage(resource: .backgroundBlue)`, потому что это безопаснее - такие ресурсы проверяются на этапе компиляции, есть автодополнение, и меньше шансов ошибиться в названии. Строки вроде `"backgroundBlueImage"` легко написать с опечаткой, и тогда изображение просто не загрузится

NaR0RoW commented on Tracker/Trackers/CreatingTracker/CreatingHabitViewController.swift:
```diff
@@ -0,0 +1,360 @@
+
+final class CreatingHabitViewController: UIViewController {
+    
+    // MARK: Pablic Property
```
Ошибка в правописании:
`Pablic` -> `Public`

Лучше включить [настройку](https://sarunw.com/posts/spell-checking-in-xcode/), чтобы `Xcode` подсвечивал ошибки в правописании

NaR0RoW commented on Tracker/Trackers/CreatingTracker/CreatingHabitViewController.swift:
```diff
@@ -0,0 +1,360 @@
+    private var scheduleDay: [WeekDay] = []
+    private var scheduleCategory = String()
+    
+    private var categoryText: String? = nil
+    private var scheduleText: String? = nil
```
`= nil` можно опустить

NaR0RoW commented on Tracker/Trackers/CreatingTracker/CreatingHabitViewController.swift:
```diff
@@ -0,0 +1,360 @@
+private extension CreatingHabitViewController {
+    
+    func getSelectedDays(from schedules: [Schedule]) -> [WeekDay] {
+        return schedules.compactMap { schedule in
+            if schedule.value, let day = schedule.day {
+                return day
+            }
+            return nil
```
Можно упростить:
`schedules.compactMap { $0.value ? $0.day : nil }`

NaR0RoW commented on Tracker/Trackers/CreatingTracker/CreatingHabitViewController.swift:
```diff
@@ -0,0 +1,360 @@
+    
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        ui.createButton.isEnabled = isSubmitButtonEnabled()
+        if ui.createButton.isEnabled {
+            ui.createButton.backgroundColor = .ypBlack
+        } else {
+            ui.createButton.backgroundColor = .ypGray
+        }
```
Лучше тернарный оператор использовать:
`ui.createButton.backgroundColor = ui.createButton.isEnabled ? .ypBlack : .ypGray`

NaR0RoW commented on Tracker/Trackers/CreatingTracker/CreatingHabitViewController.swift:
```diff
@@ -0,0 +1,360 @@
+extension CreatingHabitViewController: UITableViewDataSource {
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return isHabit ? contentsTableView.count : 1
```
В однострочных выражениях `return` можно опускать

NaR0RoW commented on Tracker/Trackers/CreatingTracker/CreatingHabitViewController.swift:
```diff
@@ -0,0 +1,360 @@
+        ui.contentsTableView.reloadData()
+        contentsCategory = categories
+        ui.createButton.isEnabled = isSubmitButtonEnabled()
+        if ui.createButton.isEnabled {
+            ui.createButton.backgroundColor = .ypBlack
+        } else {
+            ui.createButton.backgroundColor = .ypGray
+        }
```
Везде, где это возможно, лучше использовать тернарный оператор

NaR0RoW commented on Tracker/Trackers/CreatingTracker/CreatingIrregularEventViewController.swift:
Кажется файл не используется в проекте

NaR0RoW commented on Tracker/Trackers/CreatingTracker/CreatingTrackerViewController.swift:
```diff
@@ -0,0 +1,156 @@
+        let navigationController = UINavigationController(rootViewController: controller)
+        controller.delegate = self
+        navigationController.modalPresentationStyle = .popover
+        self.present(navigationController, animated: true)
```
`self` везде лучше опускать, если явно того не требуется

NaR0RoW commented on Tracker/Trackers/CustomTableViewCell/NewCategoryViewController.swift:
```diff
@@ -0,0 +1,169 @@
+    
+    func isSubmitButtonEnabled() -> Bool {
+            !(ui.newCategoryTextField.text?.isEmpty ?? false)
+    }
```
Форматирование кода немного поехало, можно выделить это место и воспользоваться автоматическим форматированием в Xcode `control + I`

NaR0RoW commented on Tracker/Trackers/View/TrackersViewController.swift:
```diff
@@ -0,0 +1,526 @@
+        datePickerValueChanged()
+    }
+    
+    func isTrackerComplitedToday(id: UUID) -> Bool {
```
`Complited` -> `Completed`

NaR0RoW commented on Tracker/Trackers/View/TrackersViewController.swift:
```diff
@@ -0,0 +1,526 @@
+        guard indexPaths.count > 0 else {
+            return nil
+        }
+        let indexPath = indexPaths[0]
```
`indexPath` никак не используется

Fennec-meow committed:
sprint_14

Changes:

1. The inability to mark the Tracker for a future date has been introduced;
2. Mutable properties have been removed from the "TrackerCategory" structure.

Fennec-meow committed:
Merge branch 'main' into sprint_14_fix
